bitcoin/BRWalletManager.c

```diff
@@ -690,17 +699,79 @@ BRWalletManagerGetWallet (BRWalletManager manager) {
+    pthread_mutex_lock (&manager->lock);
     BRSyncManagerConnect (manager->syncManager);
+    pthread_mutex_unlock (&manager->lock);
 }
 
 extern void
 BRWalletManagerDisconnect (BRWalletManager manager) {
+    pthread_mutex_lock (&manager->lock);
     BRSyncManagerDisconnect (manager->syncManager);
+    pthread_mutex_unlock (&manager->lock);
 }
 
 extern void
 BRWalletManagerScan (BRWalletManager manager) {
+    pthread_mutex_lock (&manager->lock);
     BRSyncManagerScan (manager->syncManager);
+    pthread_mutex_unlock (&manager->lock);
+}
+
+extern void
+BRWalletManagerSetMode (BRWalletManager manager, BRSyncMode mode) {
+    pthread_mutex_lock (&manager->lock);
+    if (mode != manager->mode) {
+        // get the currently known block height
+        uint64_t blockHeight = BRSyncManagerGetBlockHeight (manager->syncManager);
+
+        // kill the sync manager to prevent any additional callbacks from occuring
+        BRSyncManagerDisconnect (manager->syncManager);
+        BRSyncManagerFree (manager->syncManager);
+
+        // load up the blocks/peers saved to disk
+        BRArrayOf(BRMerkleBlock*) blocks = initialBlocksLoad(manager);
+        BRArrayOf(BRPeer) peers = initialPeersLoad(manager);
+
+        // If any of these are NULL, then there was a failure; on a failure they all need to be cleared
+        // which will cause a *FULL SYNC*
+        if (NULL == blocks || NULL == peers) {
+            if (NULL == blocks) array_new (blocks, 1);
+            else array_clear(blocks);
+
+            if (NULL == peers) array_new (peers, 1);
+            else array_clear(peers);
+        }
+
+        // set the new mode
+        manager->mode = mode;
+
+        // create the new sync manager
+        manager->syncManager = BRSyncManagerNewForMode (mode,
+                                                        manager,
+                                                        _BRWalletManagerSyncEvent,
+                                                        manager,
+                                                        (BRSyncManagerClientCallbacks) {
+                                                            _BRWalletManagerGetBlockNumber,
+                                                            _BRWalletManagerGetTransactions,
+                                                            _BRWalletManagerSubmitTransaction,
+                                                        },
+                                                        manager->chainParams,
+                                                        manager->wallet,
+                                                        manager->earliestKeyTime,
+                                                        blockHeight,
+                                                        blocks, array_count (blocks),
+                                                        peers, array_count (peers));
+    }
+    pthread_mutex_unlock (&manager->lock);
+}
```

Comment:
When moving to P2P mode this `BRSyncManagerNewForMode` might induce a full sync (from some block before `earliestKeyTime`).  BRPeerManager will examine `blocks` and if not consistent it will, I believe, discard the blocks and start anew.  Thus unless the API mode gets the blocks in place to convince/fool BRPeerManager there will, I believe, be a full sync.

Reply:
Correct. Is this an issue?

Reply:
I think the expectation would be the User can use API mode to do a fast initial sync and then switch to P2P mode.  I don't think they'd expect a full sync at that time.

Once can a) get the blocks in place to convince/fool BRPeerManager or b) change the `earliestKeyTime` to be the time of the last block.

Reply:
My understanding was that the initial use case was for P2P opting in to API, not a toggle back and forth. I'd put forth that we move ahead with this PR and track this as a separate issue as something like getting blocks will have a larger architectural impact.